contrib/internal/httptrace/httptrace.go

```diff
@@ -55,3 +81,69 @@ func FinishRequestSpan(s tracer.Span, status int, opts ...tracer.FinishOption) {
 	}
 	s.Finish(opts...)
 }
+
+// Helper function to return the IP network out of a string.
+func ippref(s string) *netaddr.IPPrefix {
+	if prefix, err := netaddr.ParseIPPrefix(s); err == nil {
+		return &prefix
+	}
+	return nil
+}
+
+// getClientIP uses the request headers to resolve the client IP. If a specific header to check is provided through
+// DD_CLIENT_IP_HEADER, then only this header is checked.
+func getClientIP(remoteAddr string, headers http.Header, clientIPHeader string) netaddr.IP {
+	ipHeaders := defaultIPHeaders
+	if len(clientIPHeader) > 0 {
+		ipHeaders = []string{clientIPHeader}
+	}
+	check := func(s string) netaddr.IP {
+		for _, ipstr := range strings.Split(s, ",") {
+			ip := parseIP(strings.TrimSpace(ipstr))
+			if !ip.IsValid() {
+				continue
+			}
+			if isGlobal(ip) {
+				return ip
+			}
+		}
+		return netaddr.IP{}
+	}
+	for _, hdr := range ipHeaders {
+		if v := headers.Get(hdr); v != "" {
+			if ip := check(v); ip.IsValid() {
+				return ip
+			}
+		}
+	}
+	if remoteIP := parseIP(remoteAddr); remoteIP.IsValid() && isGlobal(remoteIP) {
+		return remoteIP
+	}
+	return netaddr.IP{}
+}
+
+func parseIP(s string) netaddr.IP {
+	if ip, err := netaddr.ParseIP(s); err == nil {
+		return ip
+	}
+	if h, _, err := net.SplitHostPort(s); err == nil {
+		if ip, err := netaddr.ParseIP(h); err == nil {
+			return ip
+		}
+	}
+	return netaddr.IP{}
+}
+
+func isGlobal(ip netaddr.IP) bool {
+	//IsPrivate also checks for ipv6 ULA
```

Comment:
```suggestion
	// IsPrivate also checks for IPv6 ULA
```
What's IPv6 ULA and why do we care about that check? We shouldn't assume readers know.

Reply:
I'm going to quote the RFC on this one. ULA is
> an IPv6 unicast address format that is globally unique and is intended for local communications [...] They are not expected to be routable on the global Internet. They are routable inside of a more limited area such as a site.

We care to check for these addresses are not considered public, hence not global. I can always link https://www.rfc-editor.org/rfc/rfc4193.txt in the comment although I think a quick search should do the trick if someone who reads the comment gets curious